### DISCUSSION TITLE
Dual axis offsets

### DISCUSSION THREAD

**Discussion by Avatario34** [grblHAL/core#236] [Q&A] [answered]
Has anyone ever worked with the dual axis offsets $170, $171 and $172 and can confirm these are working?

I use a Teensy Board and a machine with a Dual-Y and Dual-Z setup. I need these offsets for squaring the machine. Would be great if the offset parameters worked, this would be a similar option as the M666 command in Marlin.

Second question: How do I distinguish between the two steppers per axis, is it done by positive/negative values for the $17x parameters?

Any help is greatly appreciated!

**Comment by terjeio** [comment, marked as answer]
> ...is it done by positive/negative values for the $17x parameters?

Correct, the offset is applied to the second motor.

**Reply by Avatario34** [reply]
... thanks, Terjeio. 

One further question though: To square e.g. the Y-axis, you need to move the Y-axis away from a min endstop, either at Y1 or at Y2, depeneding on the orientation of the parallelogram. This would mean that with $171 you would apply an offset to either Y1 or Y2, so to one of the two Y-axis motors. 

In Marlin this is done with positive or negative offsets: a positive offset applies to Y1, a negative to Y2. So M666=-1.5 would move the Y-axis 1.5mm away from the Y2 endstop. Is it the same logic in GRBL, i.e. $171=-1.5?

I'm asking this because you wrote 

> Correct, the offset is applied to the second motor.

If that is true, how do you move the first motor away from its endstop in case you need it?

Thanks, Christian

**Reply by terjeio** [reply]
The offset is added to the second motor after pulling off the limit switches ($27, default 1mm). Hence the pulloff distance has to be greater than the offset.

**Reply by Avatario34** [reply]
Ok, I think I‘m getting it: A negative offset moves the second motor **towards** the endstop, this is why in this case $27 needs to be bigger than that offset; it would cause a crash into the endstop otherwise, right? Will complete my controller and run some tests.

**Reply by Avatario34** [reply]
It's taken some time to build my controller and to connect the CNC, but yesterday I got it completed. 

The dual axis offsets work as expected, and I was able to square my machine in the XY plane by setting $171=1.0, ie move the y-axis at the second motor Y2 one mm away from its end stop, in addition to the pull-off distance in $27.

Now, my dual axis Z needs squaring as well, and given the physics of my machine, I need to move the **first** Z-motor Z1 0.5mm away from its endstop. My understanding of the GrblHAL logic is that I need to set a negative value to $172, ie $172=-0.5, which will actually move the **second** motor Z2 0.5mm **towards** its end stop which has the same effect as moving the first Z-motor Z1 0.5mm away from the end stop. In order to do so, I need to increase the pull-off distance in $27 by 0.5mm to give the Z2 motor enough room to move towards its end stop.

As $27 is a global setting, the above change will affect all axes and increase the pull off distances for X, Y and Z at the same time, while I only need this for the Z axis.

Is there a way to define individual pull-off distances per axis? Alternatively, could you allow separate dual axis offsets for X1 and X2 (Y and and Z respectively)? You could use the sign in $17x to determine the motor and always move this motor away from its end stop.

**Reply by terjeio** [reply]
Changing this code:

https://github.com/grblHAL/core/blob/d4f5b3b6643bde2457664ce4c3fe924331c44c87/machine_limits.c#L462-L464

to

        hal.stepper.disable_motors(auto_square, settings.axis[dual_motor_axis].dual_axis_offset < 0.0f ? SquaringMode_B : SquaringMode_A);
        if(!limits_pull_off(auto_square, fabs(settings.axis[dual_motor_axis].dual_axis_offset)))
            return false;

would work?

> Is there a way to define individual pull-off distances per axis? Alternatively, could you allow separate dual axis offsets for X1 and X2 (Y and and Z respectively)? You could use the sign in $17x to determine the motor and always move this motor away from its end stop.

Not without changing the settings data structure.

**Reply by Avatario34** [reply]
Thanks for coming back so quickly. Apologies but it is not easy for me to understand enough of what the above code is doing to judge whether it does what I was suggesting: 

If the dual_axis_offset is positive, move the corresponding "primary" (ie not ganged) motor (e.g. Y1 = Y) **away** from its endstop by dual_axis_offset. If this offset is negative, move the ganged motor (e.g. Y2 = A) **away** from its endstop by -dual_axis_offset.

You would use the sign of dual_axis_offset to decide which motor to engage, but always move the motors **away** from the endstops with positive values. By doing so, you don't need to change the global setting $27, because you always move away from the end stops. If your proposed change is intented to implement this logic, I'm more than happy to test it.

Changing the settings data structure does not make sense for this request, I'm getting it. However, my above proposal would change the semantics of the settings in $17x.

Hope this makes sense?

**Reply by terjeio** [reply]
> If the dual_axis_offset is positive, move the corresponding "primary" (ie not ganged) motor (e.g. Y1 = Y) away from its endstop by dual_axis_offset. If this offset is negative, move the ganged motor (e.g. Y2 = A) away from its endstop by -dual_axis_offset.

It is the opposite: a positive offset moves the secondary motor, negative the primary. This to keep the change backwards compatible for the secondary motor.

>  If your proposed change is intented to implement this logic, I'm more than happy to test it.

Are you able to modify the code, compile & reflash and test?

**Reply by Avatario34** [reply]
Thanks Terje, I am and will report back.

**Reply by Avatario34** [reply]
I've tested the above change, it works for my machine (Setup: dual Y axis, dual Z axis). Positive values in $17x move the Y2/Z2 motor away from the endstop, negative values move the Y1/Z1 motor away from the endstop. Exactly what I needed! Do you want me to create a PR?

**Reply by terjeio** [reply]
Thanks for the feedback. There is no need to create a PR as I have the change ready for commit. And I need to add it to the changelog.

**Reply by Avatario34** [reply]
Great, thanks Terje!

**Comment by andrewmarles** [comment]
Lots of PrintNC members use this feature.

**Reply by Avatario34** [reply]
Thank you, Andrew!
